### PR TITLE
Fix reference tracking when using nested dictionaries/objects under `py/state`

### LINKED ISSUE #500: Nested dictionaries in `py/state` incorrectly restored when tracking object references
Repro:
```python
import jsonpickle


class MyClass:
    def __init__(self, o):
        self.o = o

    def __getstate__(self):
        return self.__dict__


i1 = MyClass({"a": "a"})
i2 = MyClass({"b": "b"})
original = [i1, i1, i2, i2]

assert original[0] is original[1]  # ok
assert original[2] is original[3]  # ok

json = jsonpickle.encode(original)
# Correctly encoded as:
# [
#   {"py/object": "__main__.MyClass", "py/state": {"o": {"a": "a"}}},
#   {"py/id": 1},
#   {"py/object": "__main__.MyClass", "py/state": {"o": {"b": "b"}}},
#   {"py/id": 4}
# ]

decoded = jsonpickle.decode(json)
assert decoded[0] is decoded[1]  # ok
assert decoded[2] is decoded[3]  # fail
```

The assertion fails because `decoded[3]` is incorrectly set to the dictionary `{'a': 'a'}`, instead of the second instance of `MyClass`.

I think this only occurs for types that define `__getstate__` without also defining `__setstate__`.

This occurs because `Unpickler._restore_state` will first restore the dict under `py/state`, which tracks references of any objects under it. But then it goes into `_restore_from_dict`, which restores the same dict _again_, tracking the references of any objects under it _again_.

### PR DESCRIPTION
**Pull request by GREsau**
Fixes #500

This PR causes a dict returned from `__getstate__` (and stored in `py/state`) to no longer be restored twice, which previously caused any objects in it to be incorrectly duplicated in the Unpickler's list of object instances.

### CODE REVIEW

File: jsonpickle/unpickler.py
Comment (davvid) [inline jsonpickle/unpickler.py:629]: It looks like we're no longer calling `k = restore_key(k)` in the `else` block here. I haven't dug in deep enough to know whether or not this is an oversight.

Do all the tests still pass if we keep that?

**Comment by davvid** [comment]
Thanks for this enhancement. I'll wait until @Theelx has a chance to review. Also, we're going to want to tag and release v3.1.0 before merging this (unless Theelx thinks otherwise).

Let me know what you think about the `restore_key()` thing and then we can start planning the next release and subsequent merge.

File: jsonpickle/unpickler.py
Comment (GREsau) [reply jsonpickle/unpickler.py:629]: That is intentional to prevent double-decoding of dict keys - e.g. otherwise this would error:

```python
import jsonpickle

class Foo:
    def __init__(self):
        setattr(self, 'json://123', 'test')
    def __getstate__(self):
        return self.__dict__

f = Foo()
encoded = jsonpickle.encode(f, keys=True)
f2 = jsonpickle.decode(encoded, keys=True)
```

Comment (Theelx) [reply jsonpickle/unpickler.py:629]: This looks good to me!

File: tests/jsonpickle_test.py
Comment (Theelx) [inline tests/jsonpickle_test.py:1561]: Thanks for adding a test case!

**Review by Theelx — APPROVED** [review]
Looks fine! Sorry for not getting around to this sooner, I started a new job a few days ago, and things have been hectic.

**Comment by Theelx** [comment]
> Thanks for this enhancement. I'll wait until @Theelx has a chance to review. Also, we're going to want to tag and release v3.1.0 before merging this (unless Theelx thinks otherwise).
> 
> Let me know what you think about the `restore_key()` thing and then we can start planning the next release and subsequent merge.

This seems like something that we'd want in 3.1.0 rather than 3.1.1 (i.e. sooner rather than later). I think this could fix at least a few other open issues, so delaying it possibly a few weeks/months until 3.1.1 might not be the best. Also, if it does end up somehow accidentally breaking someone's code (unlikely), then we could still revert it and quickly release a 3.1.1 until we can figure out how to resolve it.

File: jsonpickle/unpickler.py
Comment (Theelx) [inline jsonpickle/unpickler.py:610]: Wait actually upon closer inspection, even though the method is marked private (with the single underscore), it would be best not to entirely remove this parameter unless we're in a major release (even renaming it can cause issues). Can you keep this parameter and add another one for restore_dict_items?

**Review by Theelx — CHANGES_REQUESTED** [review]
Just took another look at the code, some minor changes would be useful to preserve compatibility and add a changelog entry.

**Comment by davvid** [comment]
> This seems like something that we'd want in 3.1.0 rather than 3.1.1 (i.e. sooner rather than later).
> ...
> Also, if this PR does break something, I think it's better to have it in a minor release rather than a bugfix one

Out of extra caution I tagged v3.1.0 before this merge and I'll tag v3.2.0 now with just this merge in it after writing up a changelog entry for this one.

I should have read my older emails first ;-) Thanks for reviewing!

File: jsonpickle/unpickler.py
Comment (davvid) [reply jsonpickle/unpickler.py:610]: Shoot I missed this note. I'll tag v3.2.1 with an update that restores `ignorereserved` shortly.

**Pull request merged**
